// @fab/core/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_CONFIG_FILENAME = 'fab.config.json5';
exports.DEFAULT_MIME_TYPE = 'text/html; charset=utf-8';
exports.NO_RESPONSE_STATUS_CODE = 444;
exports.IMMUTABLE_HEADERS = {
    'cache-control': 'public, max-age=31536000, immutable',
};
exports.NON_IMMUTABLE_HEADERS = { 'cache-control': 'no-cache' };
exports.REGEXP_VALUE_PATTERN = /^\/.*\/([gimy]*)$/;
exports.ENV_VAR_SYNTAX = /^@([A-Z0-9_]+$)?/;
const CF_WORKERS_CONFIG = {
    account_id: '',
    zone_id: '',
    route: '',
    api_token: '',
    workers_dev: false,
    script_name: '',
};
const AWS_LAMBDA_CONFIG = {
    access_key: '',
    secret_key: '',
    region: '',
    cf_distribution_id: '',
    lambda_arn: '',
};
const AWS_S3_CONFIG = {
    access_key: '',
    secret_key: '',
    region: '',
    bucket_name: '',
    endpoint: '',
};
exports.HOSTING_PROVIDERS = {
    'cf-workers': {
        package_name: '@fab/deployer-cf-workers',
        capabilities: {
            server: true,
            assets: true,
        },
        config: CF_WORKERS_CONFIG,
        extension: 'js',
    },
    'aws-lambda-edge': {
        package_name: '@fab/deployer-aws-lambda',
        capabilities: {
            server: true,
            assets: false,
        },
        config: AWS_LAMBDA_CONFIG,
        extension: 'zip',
    },
    'aws-s3': {
        package_name: '@fab/deployer-aws-s3',
        capabilities: {
            server: false,
            assets: true,
        },
        config: AWS_S3_CONFIG,
        extension: 'zip',
    },
};
