// @fab/core/lib/models/ProtoFab.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const helpers_1 = require("../helpers");
const json_keys_sort_1 = __importDefault(require("json-keys-sort"));
class ProtoFab {
    constructor(init_files = {}, init_hypotheticals = {}) {
        this.files = new Map(Object.entries(init_files).map(([filename, contents]) => [
            filename,
            Buffer.from(contents),
        ]));
        this.metadata = {};
        this.hypotheticals = init_hypotheticals;
    }
    errorsPreventingCompilation() {
        // todo: an empty directory is... valid?
        const files_in_wrong_place = [];
        for (const filename of this.files.keys()) {
            if (helpers_1.filenameOutsideFabLocations(filename)) {
                files_in_wrong_place.push(filename);
            }
        }
        if (files_in_wrong_place.length > 0) {
            return `Build config leaves files outside of _assets dir: ${files_in_wrong_place.join(', ')}`;
        }
        return undefined;
    }
    serialisable() {
        const file_metadata = {};
        for (const [filename, contents] of this.files.entries()) {
            file_metadata[filename] = {
                content_type: helpers_1.getContentType(filename),
                content_length: contents.length,
            };
        }
        return {
            file_metadata,
            plugin_metadata: this.metadata,
        };
    }
    toJSON() {
        return JSON.stringify(json_keys_sort_1.default.sort(this.serialisable()));
    }
    // For testing
    _getEntries() {
        return [...this.files.entries()].map(([filename, contents]) => [
            filename,
            contents.toString('utf8'),
        ]);
    }
}
exports.ProtoFab = ProtoFab;
